src/main/resources/com/google/api/codegen/java/message.snip

```diff
@@ -0,0 +1,146 @@
+@extends "java/common.snip"
+@extends "java/method_sample.snip"
+
+@snippet generate(apiFile)
+  {@renderFileHeader(apiFile.fileHeader)}
+
+  @@Generated("by GAPIC")
+  @@BetaApi
+  public final class {@apiFile.schema.typeName} implements Serializable {
+    {@members(apiFile.schema)}
+    {@constructor(apiFile.schema)}
+    {@getters(apiFile.schema)}
+    {@newBuilder(apiFile.schema)}
+    {@builder(apiFile.schema)}
+    {@toString(apiFile.schema)}
+    {@equalsMethod(apiFile.schema)}
+    {@hashCode(apiFile.schema)}
+  }
+@end
+
+@private toString(schema)
+  @@Override
+    public String toString() {
+      return "{@schema.typeName}{"
+      @join property : schema.properties on " + \", \""
+
+          + "{@property.name}=" + {@property.name}
+      @end
+          + "}";
+  }
+
+@end
+
+@private equalsMethod(schema)
+  @@Override
+  public boolean equals(Object o) {
+    if (o == this) {
+      return true;
+    }
+    if (o instanceof {@schema.typeName}) {
+      {@schema.typeName} that = ({@schema.typeName}) o;
+      return
+      @join property : schema.properties on "&&"
+
+        this.{@property.name}.equals(that.{@property.name}())
+      @end
+      ;
+    }
+    return false;
+  }
+
+@end
+
+@private hashCode(schema)
+  @@Override
+  public int hashCode() {
+    int h = 1;
+    @join property : schema.properties
+      h *= 1000003;
```

Comment:
Is it necessary to use a different prime number? tbh this is just a copy of what AutoValue does. Java strings use 31, and Protobuffers sometimes use "1 << 16 - 1" (although each object in those examples is just an array of things of the same type).